### PR TITLE
Swan Lake alpha5 release note for runtime changes

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
$Subject

### CODE REVIEW

File: release-notes/swan-lake-alpha5-release-note.md [hunk 1]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
 
 #### New Features
 
+###### Support for configure variables with records of records, record arrays and multi dimensional array types with Config.toml.
+
+**Record of records**
```
Comment (manuranga) [inline release-notes/swan-lake-alpha5-release-note.md:52]: Shall we say `Record with record field type` or `Record with fields of type record`?

File: release-notes/swan-lake-alpha5-release-note.md [hunk 2]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+In  `Config.toml`
+
+```toml
+int2DArr = [[1,2],[3,4]]
+```
+
+###### Support for optional module name specification in TOML syntax of configurable variables.
+
+When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.
```
Comment (manuranga) [inline release-notes/swan-lake-alpha5-release-note.md:121]:
```suggestion
When providing values for the configurable variables, the  module information should be provided in `Config.toml` according to the following specifications.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 3]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```ballerina
+configurable int[][] & readonly int2DArr = ?;
+```
+In  `Config.toml`
+
+```toml
+int2DArr = [[1,2],[3,4]]
+```
+
+###### Support for optional module name specification in TOML syntax of configurable variables.
```
Comment (manuranga) [inline release-notes/swan-lake-alpha5-release-note.md:119]:
```suggestion
###### Support for optional module name in TOML syntax of configurable variables
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 1 above)
Comment (praneesha) [reply release-notes/swan-lake-alpha5-release-note.md:52]: This is too long for a heading. Is it possible to summarize and make it shorter?

E.g., "Support for configurable variables with records and arrays"?

File: release-notes/swan-lake-alpha5-release-note.md [hunk 4]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+public type County  record {
+    string name;
+};
+
+
+configurable Person person = ?;
+
+```
+
+In  `Config.toml`
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:75]:
```suggestion
The  `Config.toml` would be as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 5]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+```
+
+**Record Arrays**
+
+```ballerina
+configurable Person[] & readonly personArray = ?;
+
+```
+In  `Config.toml`
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:92]:
```suggestion
The  `Config.toml` would be as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 6]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+address.city="London"
+address.country.name="UK"
+```
+
+**Multidimensional Arrays**
+
+```ballerina
+configurable int[][] & readonly int2DArr = ?;
+```
+In  `Config.toml`
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:113]:
```suggestion
The  `Config.toml` would be as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 2 above)
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:121]:
```suggestion
When providing values for the configurable variables, the  module information can be provided in the `Config.toml` file according to the following specifications.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 7]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+
+###### Support for optional module name specification in TOML syntax of configurable variables.
+
+When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.
+
+
+* The org-name and module-name are optional for configurable variables defined in the root module of the program.
+* The org-name is optional only for configurable variables defined in the root package of the program.
+For example, consider a package with organization name `myOrg` and root module name `main`.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:126]:
```suggestion
For example, consider a package with the organization name as `myOrg` and root module name as `main`. 
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 8]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+###### Support for optional module name specification in TOML syntax of configurable variables.
+
+When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.
+
+
+* The org-name and module-name are optional for configurable variables defined in the root module of the program.
+* The org-name is optional only for configurable variables defined in the root package of the program.
+For example, consider a package with organization name `myOrg` and root module name `main`. 
+
+In `main.bal`
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:128]:
```suggestion
The  `main.bal` would be as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 9]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```ballerina
+import main.foo;
+import importedOrg/mod;
+configurable string mainVar = ?;
+public function main() {
+// use imported modules
+}
+```
+
+In `foo.bal` from module `main.foo`
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:139]:
```suggestion
The`foo.bal` file of the `main.foo` module will be as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 10]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+
+```ballerina
+configurable string fooVar = ?;
+```
+In `mod.bal` which is from another package with organization name `importedOrg` and module name `mod`,
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:145]:
```suggestion
In the `mod.bal`, which is from another package with the organization name `importedOrg` and module name `mod`, 
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 11]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```ballerina
+configurable string fooVar = ?;
+```
+In `mod.bal` which is from another package with organization name `importedOrg` and module name `mod`, 
+
+```ballerina
+configurable string modVar = ?;
+```
+
+The values can be provided in `Config.toml` as below.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:151]:
```suggestion
The values can be provided in `Config.toml` as follows.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 12]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```toml
+mainVar = "variable from root module"
+[main.foo]
+fooVar = "variable from non-root module of the root package"
+[importedOrg.mod]
+modVar = "variable from non-root package"
+```
 #### Improvements
+###### Improved command line argument parsing
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:160]:
```suggestion
###### Improved Command-Line Argument Parsing
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 13]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+The command line arguments are now parsed into:
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:162]:
```suggestion
The command-line arguments are now parsed into:
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 14]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+options
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:163]:
```suggestion
- options
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 15]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+option arguments
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:164]:
```suggestion
- option arguments
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 16]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+operands
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:165]:
```suggestion
- operands
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 17]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+**Options**
+
+Included record parameter as the last parameter specify options.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:169]:
```suggestion
Included record parameter as the last of the parameter specify options.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 18]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+public function main(*Person person) {
+	// Process data here
+}
+```
+
+```
+bal run file.bal -- --name riyafa --score=99.9
+```
+In the above example `name` and `score` are options. `riyafa` and `99.9` are option arguments.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:185]:
```suggestion
In the above example, `name` and `score` are options. `riyafa` and `99.9` are arguments of the option.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 19]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+public function main(int efficiency, string character, *Person person) {
+	// Process data here
+}
+```
+
+```
+bal run file.bal -- --name riyafa  100 --score=99.9 Good
+```
+
+This example which is the same as above includes `100` which gets mapped to `efficiency` and `Good` which gets mapped to `character` as operands.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:205]:
```suggestion
This example, which is the same as above includes `100`, which gets mapped to `efficiency`, and `Good`, which gets mapped to `character` as operands. 
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 20]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+
+```
+bal run file.bal -- --name riyafa  100 --score=99.9 Good
+```
+
+This example which is the same as above includes `100` which gets mapped to `efficiency` and `Good` which gets mapped to `character` as operands. 
+
+Both operand and option parameters can be of types int, float, decimal, string, array of any of these types and union of any of these types with nil. 
+Additionally option parameters can be of types `boolean`, `boolean[]` or `boolean?`.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:208]:
```suggestion
Additionally, option parameters can be of types `boolean`, `boolean[]`, or `boolean?`.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 21]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+**Operand arrays**
+
+Note that if there is an operand parameter of type O[], then it cannot be followed by parameters of type O[], O? and O x = d. Here O stands for a type that is a subtype of one of string,float or decimal. An array value is specified by repeatedly specifying the option parameter.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:212]:
```suggestion
>**Note:** If there is an operand parameter of type O[], then it cannot be followed by parameters of type O[], O?, and O x = d. Here O stands for a type that is a subtype of one of string, float, or decimal. An array value is specified by repeatedly specifying the `option` parameter.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 22]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+If `scores` is an int array then,
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:214]:
```suggestion
If `scores` is an int array, then,
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 23]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+```ballerina
+bal run file.bal -- --scores=10 --scores=20 --scores=30
+```
+This produces the following int array
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:219]:
```suggestion
This produces the following int array.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 24]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+[10, 20, 30]
+```
+
+**Option boolean parameters**
+
+When there’s an option of `boolean`, `boolean[]` or `boolean?` type it does not take an option argument. The presence of the option is considered to be `true` and the absence of it is considered to be false.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:226]:
```suggestion
When there’s an option of `boolean`, `boolean[]`, or `boolean?` type, it does not take an `option` argument. The presence of the option is considered to be `true` and the absence of it is considered to be false. 
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 25]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+This produces the following int array
+```
+[10, 20, 30]
+```
+
+**Option boolean parameters**
+
+When there’s an option of `boolean`, `boolean[]` or `boolean?` type it does not take an option argument. The presence of the option is considered to be `true` and the absence of it is considered to be false. 
+In the following example suppose `results` is a boolean array.
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:227]:
```suggestion
In the following example, suppose `results` is a boolean array.
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 26]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+bal run file.bal -- --results --results --results
+```
+This produces the following boolean array
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:231]:
```suggestion
This produces the following boolean array.
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 1 above)
Comment (warunalakshitha) [reply release-notes/swan-lake-alpha5-release-note.md:52]: We already have support for records and arrays. These are some complex cases related to records and arrays. Maybe I can have 3 headings.

1) Support for configurable variables with records with fields of type record
2) Support for configurable variables with arrays with fields of type record
2) Support for configurable variables with multidimensional arrays

File: release-notes/swan-lake-alpha5-release-note.md [hunk 27]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
 
 #### New Features
 
+###### Support for configure variables with records of records, record arrays and multi dimensional array types with Config.toml.
```
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:50]: ###### Support for configurable variables with records having fields of record types

File: release-notes/swan-lake-alpha5-release-note.md (hunk 1 above)
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:52]:
```suggestion
###### Support for configurable variables with records having fields of record types
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 27 above)
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:50]:
```suggestion
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 28]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```toml
+[person]
+name = "waruna"
+id = 10
+address.city="Colombo"
+address.country.name="Sri Lanka"
+
+```
+
+**Record Arrays**
```
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:86]:
```suggestion
###### Support for configurable variables with arrays having fields of record types
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 29]
```diff
@@ -47,7 +47,191 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+address.country.name="USA"
+
+[[personArray]]
+name = "hinduja"
+id = 12
+address.city="London"
+address.country.name="UK"
+```
+
+**Multidimensional Arrays**
```
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:108]:
```suggestion
###### Support for configurable variables with multidimentional arrays
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 2 above)
Comment (warunalakshitha) [inline release-notes/swan-lake-alpha5-release-note.md:121]:
```suggestion
When providing values for the configurable variables, the module information should be provided in the `Config.toml` file according to the following specifications.
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 1 above)
Comment (manuranga) [reply release-notes/swan-lake-alpha5-release-note.md]: How about `Improved support for composite configurables`

Comment (warunalakshitha) [reply release-notes/swan-lake-alpha5-release-note.md]: Then we need to consider these as improvements than new features. That should be Ok I think.

Comment (warunalakshitha) [reply release-notes/swan-lake-alpha5-release-note.md]: BTW i have removed the main heading and broken it into 3 short headings.

File: release-notes/swan-lake-alpha5-release-note.md [hunk 30]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
 
 #### New Features
 
+
+###### Support for configurable variables with records having fields of record types
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:51]:
```suggestion
###### Support for Configurable Variables with Records Having Fields of Record Types
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 31]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```toml
+[person]
+name = "waruna"
+id = 10
+address.city="Colombo"
+address.country.name="Sri Lanka"
+
+```
+
+###### Support for configurable variables with arrays having fields of record types
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:85]:
```suggestion
###### Support for Configurable Variables with Arrays Having Fields of Record Types
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 32]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+address.country.name="USA"
+
+[[personArray]]
+name = "hinduja"
+id = 12
+address.city="London"
+address.country.name="UK"
+```
+
+###### Support for configurable variables with multidimentional arrays
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:107]:
```suggestion
###### Support for Configurable Variables with Multidimentional Arrays
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 33]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```ballerina
+configurable int[][] & readonly int2DArr = ?;
+```
+The  `Config.toml` would be as follows.
+
+```toml
+int2DArr = [[1,2],[3,4]]
+```
+
+###### Support for optional module name in TOML syntax of configurable variables
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:118]:
```suggestion
###### Support for Optional Module Name in TOML Syntax of Configurable Variables
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 34]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+```
+bal run file.bal -- --name riyafa  100 --score=99.9 Good
+```
+
+This example, which is the same as above includes `100`, which gets mapped to `efficiency`, and `Good`, which gets mapped to `character` as operands. 
+
+Both operand and option parameters can be of types int, float, decimal, string, array of any of these types and union of any of these types with nil. 
+Additionally, option parameters can be of types `boolean`, `boolean[]`, or `boolean?`.
+
+**Operand arrays**
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:209]:
```suggestion
**Operand Arrays**
```

File: release-notes/swan-lake-alpha5-release-note.md [hunk 35]
```diff
@@ -47,7 +47,190 @@ To view bug fixes, see the [GitHub milestone for Swan Lake Alpha5](https://githu
...
+
+```ballerina
+bal run file.bal -- --scores=10 --scores=20 --scores=30
+```
+This produces the following int array.
+```
+[10, 20, 30]
+```
+
+**Option boolean parameters**
```
Comment (praneesha) [inline release-notes/swan-lake-alpha5-release-note.md:223]:
```suggestion
**Option Boolean Parameters**
```

File: release-notes/swan-lake-alpha5-release-note.md (hunk 1 above)
Comment (praneesha) [reply release-notes/swan-lake-alpha5-release-note.md]: +1

**Review by praneesha — APPROVED** [review]

**Pull request merged**